Design/CrashReporting/DesignDocument.md

```diff
@@ -0,0 +1,179 @@
+
+The key information that we want to capture is:
+
+* Frequency of issues
+* Types of issues
+* As much information about causes of the issues as possible
+
+Concerns that we need to bear in mind are:
+
+* Data security/privacy
+* Reducing the reporting burden on the user
+
+## Timescales
+
+We would like to have an initial version of the "Crash Reporter" in the next release of Mantid. Although it is unlikely to be possible to include all desired functionality by this time, we are confident that a level of crash reporting will be possible which should allow us to assess certain metrics, such as crash frequency. Further details are outlined in the 'Requirements' section.
+
+## Requirements
+
+#### Necessity
+
+* M : Mandatory
+* D : Desirable
+* O : Optional
+
+The "Desirable" necessity is something which might not get addressed in the
+the very first iteration of the the implementation, but should eventually be
+provided.
+
+### High Level Requirements
+
+#### Functional requirements
+
+ Label  | Requirement    |   Necessity |
+|--------|----------------|-------------|
+| R.1.1  | Uncaught exceptions are reported back as crashes  | M |
+| R.1.2  | User hard terminations i.e. via task manager are reported as crashes | M |
+| R.1.3  | Catastrophic failures are reported back as crashes | M |
+| R.1.4  | Stores crash reports in a queryable database | M|
+| R.1.5  | Where possible asks the user for further information about crash | D|
+| R.1.5  | Sends a textdump of the end of the log file along with the crash report | D|
+| R.1.6  | Obtains additional information about the crash and sends it along with crash report | D|
+
+
+#### Non-functional requirements 
+Label  | Requirement    |   Necessity |
+|--------|----------------|-------------|
+| R.3.1  | Must comply with Mantid's stated privacy policy | M |
+| R.3.2  | Must be forward compatible with the Mantid workbench | M |
+
+## Backend Design
+
+### Overview
+
+The major segments that need to be implemented in order to report crashes are: 
+* A method to detect when Mantid has crashed, it is proposed this is done within the launcher scripts.
+* A service to gather further information and then send these crashes to the web service. 
+* An interface to ask the user for additional information on these crashes. 
+* A webservice and database.  
+
+![alt text](crashdesign.png "Simple design layout")
+### Crash detection
+There are three major sorts of crashes which we wish to report with the handler.  
+
+##### 1. Uncaught exceptions
+
+ These occur when an exception propagates all the way up to the last chance error handler. Whilst it is possible to continue using Mantid after this the uncertain state of the system means that it is not recommended so it is treated like a crash.
+Detecting this sort of crash is already done as they are caught by the last chance exception handler so we just need to call whatever code we implement to report the crashes from there. 
+
+##### 2. Mantid unresponsive 
```

Comment:
I'm currently uncertain how to test the behavior here. I think what we want to test is what mantid does when it is forced to terminate by the os but can't currently think of any way to do so that is not just manually force quitting.